File: contains_duplicate_2/src/main.py
from typing import List


class Solution:
    def containsNearbyDuplicate_1(self, nums: List[int], k: int) -> bool:
        i = 0

        while i + 1 < len(nums):
            for j in range(1, k+1):
                if not i + j < len(nums):
                    break

                if nums[i] == nums[i+j]:
                    return True
            i += 1

        return False


    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        count = {}

        for i, n in enumerate(nums):
            if n in count.keys():
                count[n].append(i)
            else:
                count[n] = [i]
        
        for c in count.values():
            if len(c) < 2:
                continue

            i = 0
            for i in range(0, len(c) - 1):
                if c[i+1] - c[i] <= k:
                    return True

        return False

                
if __name__ == '__main__':
    nums = [1,2,3,1]
    k = 3
    s = Solution()
    print(s.containsNearbyDuplicate(nums, k))

File: contains_duplicate_2/src/main.rs
struct Solution {}

impl Solution {
    pub fn contains_nearby_duplicate(nums: Vec<i32>, k: i32) -> bool {
        let mut i = 0;

        while i + 1 < nums.len() {
            let mut j = 1;

            while j <= k as usize && i + j < nums.len() {
                if nums[i] == nums[i + j] {
                    return true;
                }
                j += 1;
            }
            i += 1;
        }

        false
    }
}

fn main() {
    let nums = vec![1, 2, 3, 1];
    let k = 3;
    println!("{}", Solution::contains_nearby_duplicate(nums, k));
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn tc_1() {
        let nums = vec![1, 2, 3, 1];
        let k = 3;
        assert!(Solution::contains_nearby_duplicate(nums, k));
    }

    #[test]
    fn tc_2() {
        let nums = vec![1, 0, 1, 1];
        let k = 1;
        assert!(Solution::contains_nearby_duplicate(nums, k));
    }

    #[test]
    fn tc_3() {
        let nums = vec![1, 2, 3, 1, 2, 3];
        let k = 2;
        assert!(!Solution::contains_nearby_duplicate(nums, k));
    }

    #[test]
    fn tc_4() {
        let nums = vec![99, 99];
        let k = 2;
        assert!(Solution::contains_nearby_duplicate(nums, k));
    }
}
